KAFKA-3396 : Unauthorized topics are returned to the user

edoardocomar opened the pull request:
Modified KafkaApis to return Errors.UNKNOWN_TOPIC_OR_PARTITION if
principal has no Describe access to topic

Unit tests expanded

Some paths cause the client to block due to bug
https://issues.apache.org/jira/browse/KAFKA-3727?filter=-2
tests work around this by executing in separate thread

edoardocomar commented:
Hi this work has been done with @mimaison.

Please note that we had issues with the tests because of 
https://issues.apache.org/jira/browse/KAFKA-3727 
"Consumer.poll() stuck in loop on non-existent topic manually assigned"
and
https://issues.apache.org/jira/browse/KAFKA-3728

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Modified KafkaApis to return Errors.UNKNOWN_TOPIC_OR_PARTITION if
principal has no Describe access to topic

Unit tests expanded

Some paths cause the client to block due to bug
https://issues.apache.org/jira/browse/KAFKA-3727?filter=-2
tests work around this by executing in separate thread

ijuma commented:
Can you please rebase this now that #1425 has been merged?

edoardocomar commented:
Thanks @ijuma, I did;
some tests still look ugly because of https://issues.apache.org/jira/browse/KAFKA-3727

ijuma commented:
@hachikuji Thoughts on how to handle this? It seems to me that we can't really fix this without fixing KAFKA-3727.

edoardocomar commented:
@ijuma I'd be happy to work on KAFKA-3727, but I think we need a bit of discussion on what the expected behavior should be. My comments are in that Jira issue https://issues.apache.org/jira/browse/KAFKA-3727

hachikuji commented:
@ijuma Seems we need to decide how the client should handle non-existent topics both before and after KIP-4. I can see a case for both retrying and raising an exception. I hate to suggest it, but perhaps the behavior should be configurable?

mimaison commented:
@hachikuji I agree we can probably find use cases that justify both behaviours. What surprised us mostly was the difference between assign and subscribe. 

Also for the looping behaviour, it would be nice to be able to control it with a max timeout or retries, as having to wakeup() from another thread is not very nice. In the end it sounds like a lot of new options so it might not be ideal.

hachikuji commented:
@mimaison Can you clarify what surprised about the difference between assign and subscribe?

edoardocomar commented:
@hachikuji as we wrote in https://issues.apache.org/jira/browse/KAFKA-3727

> The behavior of a consumer on poll() for a non-existing topic is **surprisingly** different/inconsistent 
> between a consumer that subscribed to the topic and one that had the topic-partition manually assigned.
> The "subscribed" consumer will return an empty collection
> The "assigned" consumer will loop forever - this feels a bug to me.

hachikuji commented:
@edoardocomar Ah, I missed that. I agree the behavior is a little surprising, but it makes sense when I think of the implementation. We currently block setting the initial position of each partition. When using subscribe(), it's not possible to be assigned a non-existing partition so this is never a problem. But obviously a manually assigned partition for a non-existent topic is a non-existing partition, and we'll block setting the position. So to fix the inconsistency, we'll have to think through the implications of removing that blocking.

edoardocomar commented:
@hachikuji @ijuma why should a missing topic or partition be a **retriable** exception 
I am thinking at least of the case where auto-create is turned off.

hachikuji commented:
@edoardocomar Sorry for the slow response. The consumer has traditionally supported the ability to be started before any producers and that means retrying on unknown topics. Even if we had some way to inspect whether autocreate has been disabled, we probably still have to support this for compatibility. I'm also thinking about the impact of the CreateTopic API in KIP-4 (which is under discussion right now). The most likely eventual outcome is that autocreation will be taken out of the broker and pushed into the producer, so it seems that the picture doesn't really change for the consumer, but we should consider whether this outcome is ideal.

One option for the problem at hand is to keep the current behavior (i.e. do nothing). Users can use the `partitionsFor()` API to verify the existence of any needed topics and raise an exception themselves if they expect it to exist ahead of time. Alternatively, we would probably need a separate configuration that lets the consumer raise an exception instead of retrying. This would call for a new KIP. Topic deletion may be an interesting case to think about as well since the `partitionsFor` workaround may not help (I'm not too sure).

All that said, I don't think we need to block this issue on a resolution to that problem. It doesn't seem like a big difference to me whether we retry indefinitely because the user is not authorized to the topic or because we are waiting for a topic that may never be created. In either case, the user ultimately has to monitor the progress of the application and take steps when it is failing or falling behind.

ijuma commented:
My concern is that configuring security is challenging enough as it is. Configuring authorization incorrectly is very common and harder to diagnose than a topic that hasn't been created (even the latter sometimes causes people to be confused and waste time).

Replacing an exception with indefinite blocking is making the user experience during development even worse than it is at the moment (which is already suboptimal in a number of situations). Ideally, we would be moving in the other direction.

If we decide to do this, then at the very least, we need to think about how users will diagnose the problem and update the authorization documentation with the information.

hachikuji commented:
@ijuma At the very least, we can change the consumer/producer log messages to warn the user that the topic might not exist or they might not have access to it. I agree this is less than ideal, but I don't see an alternative other than what I suggested above (changing how the client handles unknown topics). That is, unless we think this patch is not worth having at all. I was a little torn on this initially, but it does seem better to avoid leaking topic information.

ijuma commented:
@hachikuji The log warning you have suggested seems like our best bet at the moment.

edoardocomar commented:
Hi @ijuma @hachikuji 
we have made a PR https://github.com/apache/kafka/pull/1535 
to address the non existent topic or partition issue https://issues.apache.org/jira/browse/KAFKA-3727

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -701,22 +707,29 @@ class KafkaApis(val requestChannel: RequestChannel,
         java.util.Collections.emptyList()))
 
+    // do not disclose the existence of the unauthorized topic    
+    val unauthorizedTopicMetadata = unauthorizedTopics.map(topic =>
```
I think we need to handle the all topics case specially:

``` scala
 val topics =
       // Handle old metadata request logic. Version 0 has no way to specify "no topics".
       if (requestVersion == 0) {
         if (metadataRequest.topics() == null || metadataRequest.topics().isEmpty)
           metadataCache.getAllTopics()
         else
           metadataRequest.topics.asScala.toSet
       } else {
         if (metadataRequest.isAllTopics)
           metadataCache.getAllTopics()
         else
           metadataRequest.topics.asScala.toSet
       }
```

For that case, we want to filter the topics that the user doesn't have DESCRIBE access to.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -228,12 +228,17 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val (authorizedRequestInfo, unauthorizedRequestInfo) = filteredRequestInfo.partition {
+      val filteredRequestInfo1 = offsetCommitRequest.offsetData.asScala.toMap -- invalidRequestsInfo.keys
+      
+      //filter topics with Describe ACL as if they were non-existent
```
Maybe `treat topics without Describe ACL as if they were non-existent`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -244,7 +249,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
-        val combinedCommitStatus = mergedCommitStatus.mapValues(new JShort(_)) ++ invalidRequestsInfo.map(_._1 -> new JShort(Errors.UNKNOWN_TOPIC_OR_PARTITION.code))
+        var combinedCommitStatus = mergedCommitStatus.mapValues(new JShort(_)) ++ invalidRequestsInfo.map(_._1 -> new JShort(Errors.UNKNOWN_TOPIC_OR_PARTITION.code)) 
+        combinedCommitStatus = combinedCommitStatus ++ unauthorizedForDescribeRequestInfo.mapValues(_ => Errors.UNKNOWN_TOPIC_OR_PARTITION.code).mapValues(new JShort(_))
```
We don't need the var, just `++` to the previous one right?

ijuma replied:
Something like the following maybe (I didn't try to compile it):

``` scala
val unknownTopicStatus = (invalidRequestsInfo ++ unauthorizedForDescribeRequestInfo).mapValues(_ => Errors.UNKNOWN_TOPIC_OR_PARTITION.code)
val combinedCommitStatus = (mergedCommitStatus ++ unknownTopicStatus).mapValues(new JShort(_))
```

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -244,7 +249,8 @@ class KafkaApis(val requestChannel: RequestChannel,
               s"on partition $topicPartition failed due to ${Errors.forCode(errorCode).exceptionName}")
```
We should also log information about the unauthorized for describe with debug enabled, probably.

hachikuji replied:
I'm not sure, but it seems unintentional. I can't think of a reason why we wouldn't want to log the errors.

ijuma replied:
Just to check: are you going to address my comment with regards to adding additional text for the unauthorized for describe case when logging? It would be helpful to have that information on the server-side logs when debugging issues. It is possible to correlate with the authorizer logs, but that's more work.

edoardocomar replied:
logged

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -228,12 +228,17 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val filteredRequestInfo = offsetCommitRequest.offsetData.asScala.toMap -- invalidRequestsInfo.keys
-
-      val (authorizedRequestInfo, unauthorizedRequestInfo) = filteredRequestInfo.partition {
+      val filteredRequestInfo1 = offsetCommitRequest.offsetData.asScala.toMap -- invalidRequestsInfo.keys
```
I think I would name this `existingTopicsRequestsInfo`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -228,12 +228,17 @@ class KafkaApis(val requestChannel: RequestChannel,
       val invalidRequestsInfo = offsetCommitRequest.offsetData.asScala.filter { case (topicPartition, _) =>
```
We can use `partition` here too to simplify the code. And the variable should probably be named `nonExistingRequestsInfo`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -228,12 +228,17 @@ class KafkaApis(val requestChannel: RequestChannel,
+      val filteredRequestInfo1 = offsetCommitRequest.offsetData.asScala.toMap -- invalidRequestsInfo.keys
+      
+      //filter topics with Describe ACL as if they were non-existent
+      val (filteredRequestInfo2, unauthorizedForDescribeRequestInfo) = filteredRequestInfo1.partition {
```
And then this becomes `filteredRequestInfo`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -701,22 +707,29 @@ class KafkaApis(val requestChannel: RequestChannel,
     var (authorizedTopics, unauthorizedTopics) =
       topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
 
+    var uncreatableTopics = Set[String]()
```
Maybe `unauthorizedForCreateTopics`

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -701,22 +707,29 @@ class KafkaApis(val requestChannel: RequestChannel,
     var (authorizedTopics, unauthorizedTopics) =
```
Maybe `unauthorizedForDescribeTopics`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -762,8 +775,6 @@ class KafkaApis(val requestChannel: RequestChannel,
         authorize(request.session, Describe, new Resource(Topic, topicPartition.topic))
       }
-      val unauthorizedTopicResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.TOPIC_AUTHORIZATION_FAILED.code)
-      val unauthorizedStatus = unauthorizedTopicPartitions.map(topicPartition => (topicPartition, unauthorizedTopicResponse)).toMap
```
You are not returning any data for the unauthorized topics and you should return `UNKNOWN_TOPIC_OR_PARTITION`, right? Are there tests missing then?

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -353,7 +353,9 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
-      case e: TopicAuthorizationException => assertEquals(Collections.singleton(topic), e.unauthorizedTopics());
+      case e: TestTimeoutException => //expected
+    } finally {
+      this.consumers.foreach(_.wakeup())
```
Why do we have this?

ijuma replied:
You explained this in another test, probably worth having the same comment here.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -433,7 +435,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
   }
 
-  @Test(expected = classOf[TopicAuthorizationException])
+  @Test(expected = classOf[KafkaException])
```
Isn't there a more specific exception?

hachikuji replied:
See my comment above. I would like to make the current behavior more explicitly defined.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -479,11 +481,35 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
 
-  @Test(expected = classOf[TopicAuthorizationException])
+  @Test(expected = classOf[TestTimeoutException])
   def testOffsetFetchWithNoTopicAccess() {
```
Can we please improve the test names to be more specific? For example, `testOffsetFetchWithNoDescribeTopicAccess` (if applicable).

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -203,23 +225,68 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+    AclCommand.main(describeAclArgs)
+    servers.foreach(s => {
+      TestUtils.waitAndVerifyAcls(TopicDescribeAcl, s.apis.authorizer.get, topicResource)
+    })
```
Nitpick:

``` scala
servers.foreach { s =>
  TestUtils.waitAndVerifyAcls(TopicDescribeAcl, s.apis.authorizer.get, topicResource)
}
```

ijuma replied:
There is at least one other case like this.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -203,23 +225,68 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+      consumeRecords(this.consumers.head)
+      fail("exception expected")
+    } catch {
+       case e: TestTimeoutException => //expected from consumeRecords()
```
Why do we specify `from consumeRecords()` given that it's the only method called in the try?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -284,15 +395,35 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
-      iters += 1
     }
+    t.start()
+    t.join(10000L)
```
I suggest you use scala Futures for this, the code should be a lot simpler. We should probably synchronize access to the consumer since it's not thread-safe though.

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -302,3 +433,7 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
 }
+
+class TestTimeoutException (msg:String)
+  extends java.lang.RuntimeException (msg) {}
```
This should be simply:

``` scala
class TestTimeoutException(msg: String) extends java.lang.RuntimeException(msg)
```

It seems like we use this in another test too, right? If so, we should move it to its own file.

ijuma commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -475,9 +475,11 @@ object TestUtils extends Logging {
     producerProps.put(ProducerConfig.ACKS_CONFIG, acks.toString)
-    producerProps.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, maxBlockMs.toString)
+    if(!producerProps.containsKey(ProducerConfig.MAX_BLOCK_MS_CONFIG))
+       producerProps.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, maxBlockMs.toString)
```
It's not good to introduce this inconsistency here. Can you not just pass the parameter via the `createNewProducer` call?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -229,16 +296,60 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
     sendRecords(numRecords, tp)
-    //Consume records
+
+    //Deleting topic ACL without asking for console confirmation
```
Why are you mentioning `without asking for console confirmation`? Isn't that the case for all ACLs calls here?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -203,23 +225,68 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+      fail("exception expected")
     } catch {
-      case e: TopicAuthorizationException => //expected
+      case e: org.apache.kafka.common.errors.TimeoutException => //expected
```
Why are we using the fully qualified name here?

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -597,21 +624,42 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
+
+    val t = new java.lang.Thread() {
+      override def run () {
+        println("in consumeRecords.Thread.run")
```
We don't commit `println`s in tests.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -597,21 +624,42 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
+      throw caughtExc
+    
+    if (records.isEmpty()) 
+        throw new TestTimeoutException("Failed to consume the expected records after 10000 millis.")
```
It sounds like we should have a helper test method for this functionality as there is similar code in another test (maybe in `TestUtils`). And as I said elsewhere, please use scala Futures.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -544,7 +568,10 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
-    val possibleErrorCodes = resources.map(_.errorCode)
+    var possibleErrorCodes = resources.map(_.errorCode)
+    if (key == ApiKeys.METADATA.id) 
+      possibleErrorCodes = possibleErrorCodes + Errors.UNKNOWN_TOPIC_OR_PARTITION.code()
```
Looks like `ResourceType.errorCode` is lying. Not sure if that method makes sense, maybe we should remove it since it's only used by tests, but that needs further discussion so let's just ignore it for this PR.

Is `METADATA` the only case where we can also have `UNKNOWN_TOPIC_OR_PARTITION`? I thought there were more of them based on the code changes.

ijuma commented:
I've been thinking about this and I think we should go ahead with this and then look at ways at improving it further once more of the KIP-4 work is in.

I left a few review comments. In addition to that, we should make the change to the log message as per Jason's suggestion.

Also, users can use `listTopics()` or `partitionsFor()` that don't hang as a workaround (Jason's suggestion).

mimaison commented:
Thanks for the review, I'll update the PR.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -265,8 +266,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
     } catch {
-      case e: TopicAuthorizationException =>
-        assertEquals(Collections.singleton(topic), e.unauthorizedTopics())
+      case e: org.apache.kafka.common.errors.TimeoutException => //expected
```
Why are we using the fully qualified name?

ijuma commented:
@mimaison, do you think you'll have a chance to look at this sometime soon?

mimaison commented:
Sorry for the delay, yes I'm working on it today

ijuma commented:
@mimaison no worries about the delay and there's no rush. :) I was just trying to get a rough idea about the timing. Thanks!

mimaison commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -433,7 +435,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
   }
 
-  @Test(expected = classOf[TopicAuthorizationException])
+  @Test(expected = classOf[KafkaException])
```
In this case, we don't have DESCRIBE on the topic so the broker returns UNKNOWN_TOPIC_OR_PARTITION and in OffsetCommitResponseHandler.handle() this error is not thrown as a KafkaException. Should we instead raise it as is ?

ijuma replied:
It seems to me that it would be better. @hachikuji, what do you think?

mimaison replied:
Because UToP is retriable, it's not simple to make it pop back up. In this case, if the raise it, we end up stuck in the while(true) loop in ConsumerCoordinator.commitOffsetsSync(). Basically we get back in the problematic we raised in https://issues.apache.org/jira/browse/KAFKA-3727 where we didn't reach a consensus. 
I'll be inclined to keep it like this for now and try to deal with it in 3727.

hachikuji replied:
I think an UNKNOWN_TOPIC_OR_PARTITION error in commit can probably be treated as a fatal error which we raise to the user. If the topic doesn't exist, then you shouldn't have any offsets to commit anyway. Maybe we can do the following:
1. Update the possible error codes in `OffsetCommitResponse` and the protocol documentation to include this error code.
2. Catch the error explicitly in `OffsetCommitResponseHandler` and rethrow either as a `UnknownTopicOrPartitionException` with a better message or `KafkaException` as is currently done. At least then the expected behavior is clearly implemented.

I'm wondering if it makes sense to do similarly for the OffsetFetch API. We may actually prefer in that case to retry since the topic may be created automatically by a pending TopicMetadata request. That said, it may be better to keep the current behavior for the sake of this patch. Also, since no one is complaining about it that I've seen, maybe we shouldn't "fix" it to be consistent with behavior that we want to deprecate anyway. In that case, we should do 1. and 2. similarly for offset fetches. Thoughts?

ijuma replied:
Your suggestion sounds good to me @hachikuji

mimaison replied:
Let's see how KIP-4 will change things and we'll carry on the discussion in KAFKA-3727

mimaison committed:
KAFKA-3396: Updates and cleanups following the feedback

mimaison commented:
I've updated the PR and rebased on trunk

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -231,11 +231,15 @@ class KafkaApis(val requestChannel: RequestChannel,
+        case (topicPartition, _) => !metadataCache.contains(topicPartition.topic)
+      }
+
+      // treat topics with Describe ACL as if they were non-existent
```
`with` should be `without`, right?

hachikuji replied:
nitpick: should that be "without"?

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -244,7 +249,8 @@ class KafkaApis(val requestChannel: RequestChannel,
               s"on partition $topicPartition failed due to ${Errors.forCode(errorCode).exceptionName}")
```
Thanks for updating the code so that these errors are logged as per my comment above. I thought that we should indicate in the server log that the issue is lack of describe instead of a missing topic. Something as simple as some extra text at the end: "as the user u does not have DESCRIBE permission" (something like that anyway).

ijuma replied:
Also, the previous code did not include non existing topics from the server log. Not sure if that was intentional. @hachikuji, do you know?

ijuma commented:
@mimaison, it would be good to port the tests from https://github.com/apache/kafka/commit/44ad3ec77364fab869cb5e01a182341bd0ae2698 to make sure the behaviour of pattern subscriptions is correct.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -231,11 +231,15 @@ class KafkaApis(val requestChannel: RequestChannel,
+      }
+
+      // treat topics with Describe ACL as if they were non-existent
+      val (filteredRequestInfo, unauthorizedForDescribeRequestInfo) = existingTopicsRequestsInfo.partition {
```
Seems like it would be possible to merge these first two partition calls. Something like this:

``` scala
val (existingTopicInfo, nonExistingTopicInfo) = offsetCommitRequest.offsetData.asScala.toMap.partition {
  case (topicPartition, _) => authorize(request.session, Describe, new Resource(Topic, topicPartition.topic)) && 
     metadataCache.contains(topicPartition.topic)
}
```

That might also make the expectation that you need Describe access to check topic existence clearer. Thoughts?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -705,25 +709,36 @@ class KafkaApis(val requestChannel: RequestChannel,
     if (authorizedTopics.nonEmpty) {
       val nonExistingTopics = metadataCache.getNonExistingTopics(authorizedTopics)
       if (config.autoCreateTopicsEnable && nonExistingTopics.nonEmpty) {
         authorizer.foreach { az =>
```
Not part of this patch, but the `authorize` function seems to already handle this, so could we replace this and the `if` below with a call to `authorize(request.session, Create, Resource.ClusterResource)`?

hachikuji replied:
Can you fix this please? Is there any reason not to?

edoardocomar replied:
done thanks

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -705,25 +709,36 @@ class KafkaApis(val requestChannel: RequestChannel,
+        Set.empty[MetadataResponse.TopicMetadata]
+      else
+        unauthorizedForDescribeTopics.map(topic =>
+          new MetadataResponse.TopicMetadata(Errors.UNKNOWN_TOPIC_OR_PARTITION, topic, common.Topic.isInternal(topic), java.util.Collections.emptyList()))
```
Should we reveal whether the topic is internal if the user has no Describe access?

hachikuji replied:
Can you address this comment?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -705,25 +709,36 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
 
-    val unauthorizedTopicMetadata = unauthorizedTopics.map(topic =>
+    val uncreatableTopicMetadata = unauthorizedForCreateTopics.map(topic =>
```
Would it make sense to change this to `unauthorizedForCreateTopicMetadata`? Similarly, we could rename the variable below to `unauthorizedForDescribeTopicMetadata`.

hachikuji commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -352,7 +356,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
 
   @Test
-  def testConsumeWithNoTopicAccess() {
+  def testConsumeWithoutTopicDescribeAccess() {
```
The expectation of this test case is not too clear. It looks like the exception that we raise is actually being thrown from the offset fetch request, and it seems like it may be a separate bug that we don't handle the UNKNOWN_TOPIC_OR_PARTITION error code in the offset fetch response handler. Maybe we should fix that? Ultimately, until we have timeouts for some of the consumer's blocking calls, we may not have a great way to test this case.

hachikuji replied:
Yes, but we can make the current behavior explicit, right? This test only passes because we do not handle the UNKNOWN_TOPIC_OR_PARTITION error in the offset fetch response handler. So:
1. In `ConsumerCoordinator.OffsetFetchResponseHnadler`, let's add an explicit case for the `UNKNOWN_TOPIC_OR_PARTITION`. I suggest we let the error be raised as an instance of `UnknownTopicOrPartitionException` instead of a generic `KafkaException`. We should also give the user a message which lets them know that they may not have Describe access on the topic.
2. Let's do the same for `ConsumerCoordinator.OffsetCommitResponseHandler`.
3. Let's change the name of this test case so that it's clear what exactly is expected to fail.

edoardocomar replied:
Hi 
sorry but could you confirm ... :
If `ConsumerCoordinator.OffsetFetchResponseHandler` raises an UTOP 
that is a retriable exception and `ConsumerCoordinator.fetchCommittedOffsets` remains stuck in the `while` loop.
The client would not see the exception. The unit test would have to be coded for a timeout. 
Is this the behavior you'd like to see ? With a log entry while stuck in the loop ?

hachikuji replied:
@edoardocomar Hmm... Yeah, that's a fair point. How about we catch the error explicitly and throw as `KafkaException` with a message which indicates the topic/partition is unknown and mentions that the the user may not have describe access?

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -705,25 +709,36 @@ class KafkaApis(val requestChannel: RequestChannel,
+        Set.empty[MetadataResponse.TopicMetadata]
+      else
+        unauthorizedForDescribeTopics.map(topic =>
+          new MetadataResponse.TopicMetadata(Errors.UNKNOWN_TOPIC_OR_PARTITION, topic, common.Topic.isInternal(topic), java.util.Collections.emptyList()))
```
Good point, we should always return `false` here.

mimaison committed:
KAFKA-3396: More updates + rebased off trunk

- Added tests from 44ad3ec
- Small refactorings

mimaison commented:
Thanks for all the feedback ! I've updated the PR

mimaison commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -352,7 +356,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
 
   @Test
-  def testConsumeWithNoTopicAccess() {
+  def testConsumeWithoutTopicDescribeAccess() {
```
Yes it's the same issue we've discussed above. Until we decide how we want to handle OTOP, we'll have some ugly code.

mimaison commented:
@ijuma @hachikuji I believe we've addressed all the comments, can you have another look ?

ijuma commented:
Sorry for the delay @mimaison and thanks for the reminder. Will try to look at this soon.

ijuma commented:
@mimaison, with the feature freeze in place, this is one of the main bugs we want to get in for 0.10.1.0. Could you please merge trunk into this branch? KIP-79 made some changes to the list offsets request so it may need some updates. Sorry about that. :(

edoardocomar commented:
@ijuma @hachikuji - in hindsight, shouldn't we also handle an offsetRequest not to return TOPIC_AUTHORIZATION_FAILED where Describe auth is checked. e.g. 

```
private def handleOffsetRequestV0(request : RequestChannel.Request) ... {
...
    val (authorizedRequestInfo, unauthorizedRequestInfo) = offsetRequest.offsetData.asScala.partition {
      case (topicPartition, _) => authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
    }

    val unauthorizedResponseStatus = unauthorizedRequestInfo.mapValues(_ =>
      new PartitionData(Errors.TOPIC_AUTHORIZATION_FAILED.code, List[JLong]().asJava)
    )
```

it seems we had missed it

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Rebased after kip-79 changes.
Fixing leak of topic for LIST_OFFSETS when unauthorized.
Added tests.

edoardocomar commented:
@ijuma @hachikuji the Jenkins build failure 
https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/5818/
looks unrelated to our changes. Can you please review ? thanks

I'm forcing a new rebuild by rebasing on trunk rather than 0.10.1

ijuma commented:
@edoardocomar, we should, yes. I think that was "missed" because that code was moved as part of KIP-79, but I haven't checked in detail yet. Will check the PR today or tomorrow. Thanks!

edoardocomar commented:
@ijuma Jenkins has failed with a timeout, the tests locally pass for me can you please kick Jenkins again ? thanks

edoardocomar commented:
@ijuma @hachikuji can you please take a look at the revised PR? thanks

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -234,38 +234,37 @@ class KafkaApis(val requestChannel: RequestChannel,
-      // filter non-existent topics
-      val invalidRequestsInfo = offsetCommitRequest.offsetData.asScala.filter { case (topicPartition, _) =>
-        !metadataCache.contains(topicPartition.topic)
+
```
nit: unneeded newline?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -234,38 +234,37 @@ class KafkaApis(val requestChannel: RequestChannel,
-        !metadataCache.contains(topicPartition.topic)
+
+      val (existingOrAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) = offsetCommitRequest.offsetData.asScala.toMap.partition {
+        case (topicPartition, _) => metadataCache.contains(topicPartition.topic) && authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
```
nit: not much difference here, but it seems more natural to check authorization first.

ijuma replied:
`MetadataCache` has two methods that do the same thing: `contains` and `hasTopicMetadata`. We should remove one of them to avoid confusion. Also, we call `hasTopicMetadata` in the block for `if (header.apiVersion == 0)`, which seems unnecessary as we have already done it here (it doesn't prevent potential race conditions either although it does narrow the window a little).

edoardocomar replied:
ok

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -811,25 +810,36 @@ class KafkaApis(val requestChannel: RequestChannel,
+      new MetadataResponse.TopicMetadata(Errors.TOPIC_AUTHORIZATION_FAILED, topic, common.Topic.isInternal(topic),
         java.util.Collections.emptyList()))
 
+    // do not disclose the existence of unauthorized topics
```
nit: should be obvious, but we may as well emphasize unauthorized _for Describe_.

edoardocomar replied:
thanks

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -811,25 +810,36 @@ class KafkaApis(val requestChannel: RequestChannel,
+    // do not disclose the existence of unauthorized topics
+    val unauthorizedForDescribeTopicMetadata =
+      // In case of all topics, don't include unauthorized topics
+      if ((requestVersion == 0 && (metadataRequest.topics() == null || metadataRequest.topics().isEmpty)) || (metadataRequest.isAllTopics))
```
nit: Doesn't seem like the parenthesis around `metadataRequest.isAllTopics` are necessary.

hachikuji commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -591,9 +697,14 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+      if (resourceType == Topic)
+        if (apiKey == ApiKeys.PRODUCE || apiKey == ApiKeys.FETCH ) {
+          //Only allowing TOPIC_AUTHORIZATION_FAILED as an error code
+          Seq(resourceType.errorCode)
```
Might be clearer to reference the TOPIC_AUTHORIZATION error code directly?

edoardocomar replied:
ok

ijuma replied:
I think this is not right. Both fetch and produce should also return `UNKNOWN_TOPIC_OR_PARTITION` if given a topic that the user doesn't have DESCRIBE permission for, right? It seems like neither checks if the user has DESCRIBE permission, which seems to defeat the point of this PR.

edoardocomar replied:
addressed thanks

hachikuji commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -635,15 +746,21 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
-      for (record <- consumer.poll(50).asScala) {
-        records.add(record)
+
+    val future = Future {
```
It's a little weird here to bound this loop by the number of iterations. Could we refactor it to use `waitUntil`? For example:

``` scala
    TestUtils.waitUntilTrue(() => {
        for (record <- consumer.poll(50).asScala) 
          records.add(record)
        records.size >= numRecords
    })
```

Either that or use the timeout directly in the loop instead of `maxIters`?

edoardocomar replied:
thanks

hachikuji commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -203,42 +229,127 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
 
-  /**
+  @Test
+  def testNoProduceAclWithDescribeAcl {
```
nit: The first "Acl" seems out of place. Should this be `testNoProduceWithDescribeAcl`? A few more like this below.

edoardocomar replied:
thanks

hachikuji commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -203,42 +229,127 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+    }
+  }
+  
+  private def noConsumeAclWithDescribeAclProlog {
```
nit: What is "Prolog"?

edoardocomar replied:
a Setup step. renamed

hachikuji commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -457,7 +457,6 @@ object TestUtils extends Logging {
     val producerProps = props.getOrElse(new Properties)
     producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
     producerProps.put(ProducerConfig.ACKS_CONFIG, acks.toString)
-    producerProps.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, maxBlockMs.toString)
```
Why was this removed?

edoardocomar replied:
To be able to set it to a low value in EndToEndAuthorizationTest 
to allow some test to finish quickly

ijuma replied:
This change causes the parameter of `createNewProducer` to be ignored affecting a bunch of other tests though. It seems like the easiest thing is to add a couple of protected methods in `IntegrationTestHarness`: `createProducer` and `createConsumer`. You can then override that in `EndToEndAuthorizationTest`.

edoardocomar replied:
good point, done

hachikuji commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -284,21 +395,30 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
-      for (record <- consumer.poll(50).asScala) {
-        records.add(record)
+    
+    val future = Future {
```
Same as above. Seems like we could make this a more conventional timeout loop.

edoardocomar replied:
ok

hachikuji reviewed (commented):
Thanks for the patch! I left some comments (mostly nitpicks). I do think it's worthwhile treating UNKNOWN_TOPIC_OR_PARTITION explicitly in OffsetFetch and OffsetCommit response handlers on the client though.

mimaison commented:
Thanks for the feedback, we'll update the PR on Monday.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -234,38 +234,37 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val invalidRequestsInfo = offsetCommitRequest.offsetData.asScala.filter { case (topicPartition, _) =>
-        !metadataCache.contains(topicPartition.topic)
+
+      val (existingOrAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) = offsetCommitRequest.offsetData.asScala.toMap.partition {
```
The first variable should be `existingAndAuthorizedForDescribeTopics`.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -234,38 +234,37 @@ class KafkaApis(val requestChannel: RequestChannel,
       // the callback for sending an offset commit response
       def sendResponseCallback(commitStatus: immutable.Map[TopicPartition, Short]) {
-        val mergedCommitStatus = commitStatus ++ unauthorizedRequestInfo.mapValues(_ => Errors.TOPIC_AUTHORIZATION_FAILED.code)
+        var combinedCommitStatus = commitStatus.mapValues(new JShort(_)) ++
```
Should be a val.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -438,6 +469,83 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), groupResource)
+    this.consumers.head.subscribe(Pattern.compile(topicPattern), new NoOpConsumerRebalanceListener)
+    this.consumers.head.poll(50)
+    assertTrue(this.consumers.head.subscription().isEmpty())
```
No need for `()` in `subscription` and `isEmpty`.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -438,6 +469,83 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
   }
 
   @Test
+  def testPatternSubscriptionWithNoTopicAccess() {
```
We should extend the pattern subscription tests to also include the case where the user has DESCRIBE, but not READ permission. In those cases, we should still get `TopicAuthorizationException`.

edoardocomar replied:
done thanks

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -438,6 +469,83 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    // set the subscription pattern to an internal topic that the consumer has no read permission for, but since
+    // `exclude.internal.topics` is true by default, the subscription should be empty and no authorization exception
+    // should be thrown
+    consumer.subscribe(Pattern.compile(kafka.common.Topic.GroupMetadataTopicName), new NoOpConsumerRebalanceListener)
```
Given the implementation in `trunk`, this comment isn't really helpful anymore. I would add an additional case where the consumer has DESCRIBE, but no READ permission for the internal topic. And then move this comment there (and tweak it slightly).

edoardocomar replied:
test added

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -438,6 +469,83 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+      consumer.subscribe(Pattern.compile(".*"), new NoOpConsumerRebalanceListener)
+      consumeRecords(consumer)
+    } catch {
+      case e: TestTimeoutException => //expected
```
Why do we get this instead of an empty subscription?

edoardocomar replied:
oops, fixed test

ijuma reviewed (commented):
Thanks for your patience. I left a few more comments. The most important ones are related to how we don't seem to check `DESCRIBE` permissions for `fetch` and `produce`.

edoardocomar commented:
@ijuma We have handled the FETCH and PRODUCE commands

edoardocomar commented:
As I have squashed some commits, I am opening  a new PR
